### PR TITLE
Correct cardinality in join output

### LINKED ISSUE #20315: abort in debug: LowCardinality -> Nullable -> Vector
https://clickhouse-test-reports.s3.yandex.net/20224/15ce3dc1124445c1edd7dde26ca998ed67816150/fuzzer_debug/report.html#fail1

```
2021.02.09 13:44:25.975671 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Debug> executeQuery: (from [::1]:42050, using production parser) SELECT 1023 + l, * FROM (SELECT toLowCardinality(toNullable(number)) AS l FROM system.numbers LIMIT 10) AS s1 ANY LEFT JOIN (SELECT toLowCardinality(toNullable(number)) AS r FROM system.numbers LIMIT 7) AS s2 ON (l + 1023) = (r * 3)
2021.02.09 13:44:25.978451 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> ContextAccess (default): Access granted: SELECT(number) ON system.numbers
2021.02.09 13:44:25.980586 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> ContextAccess (default): Access granted: SELECT(number) ON system.numbers
2021.02.09 13:44:25.983013 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> ContextAccess (default): Access granted: SELECT(number) ON system.numbers
2021.02.09 13:44:25.987689 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> ContextAccess (default): Access granted: SELECT(number) ON system.numbers
2021.02.09 13:44:25.988498 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2021.02.09 13:44:25.992361 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2021.02.09 13:44:25.993567 [ 91 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> InterpreterSelectQuery: FetchColumns -> Complete
2021.02.09 13:44:26.005592 [ 162 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> CreatingSetsTransform: Creating join.
2021.02.09 13:44:26.005993 [ 162 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Debug> CreatingSetsTransform: Created Join with 7 entries from 7 rows in 0.000365869 sec.
2021.02.09 13:44:26.010925 [ 162 ] {db694e24-a7a9-4729-a1c2-90ccb6672069} <Trace> PipelineExecutor: Thread finished. Total time: 0.005616319 sec. Execution time: 0.004462349 sec. Processing time: 0.001097067 sec. Wait time: 5.6903e-05 sec.
2021.02.09 13:44:26.011292 [ 168 ] {} <Fatal> BaseDaemon: ########################################
2021.02.09 13:44:26.011718 [ 168 ] {} <Fatal> BaseDaemon: (version 21.3.1.5957, build id: 8DD631BE1F9992A50D94565D97F7EA3FBDDCB102) (from thread 91) (query_id: db694e24-a7a9-4729-a1c2-90ccb6672069) Received signal Aborted (6)
2021.02.09 13:44:26.011942 [ 168 ] {} <Fatal> BaseDaemon: 
2021.02.09 13:44:26.012144 [ 168 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f766a04518b 0x7f766a024859 0x7f766a024729 0x7f766a035f36 0x10e7c036 0x1a4b9e53 0x1a4743ec 0x1a4742a3 0x1a42ef66 0x1a410aee 0x1a410962 0x1a410880 0x196db6ff 0x197a05e0 0x1a0d11fc 0x1a0d1647 0x1ac7a0cc 0x1ac7898d 0x1ac728e8 0x1ac7ea78 0x1e4fb39c 0x1e4fbba0 0x1e638d13 0x1e635b9d 0x1e634a28 0x7f766a1fa609 0x7f766a121293
2021.02.09 13:44:26.012411 [ 168 ] {} <Fatal> BaseDaemon: 4. raise @ 0x4618b in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.02.09 13:44:26.012596 [ 168 ] {} <Fatal> BaseDaemon: 5. abort @ 0x25859 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.02.09 13:44:26.012736 [ 168 ] {} <Fatal> BaseDaemon: 6. ? @ 0x25729 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.02.09 13:44:26.012908 [ 168 ] {} <Fatal> BaseDaemon: 7. ? @ 0x36f36 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.02.09 13:44:26.014450 [ 168 ] {} <Fatal> BaseDaemon: 8. ./obj-x86_64-linux-gnu/../src/Common/PODArray.h:0: DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 15ul, 16ul>::operator[](long) const @ 0x10e7c036 in /workspace/clickhouse
2021.02.09 13:44:26.067013 [ 168 ] {} <Fatal> BaseDaemon: 9. ./obj-x86_64-linux-gnu/../src/Columns/ColumnVector.h:343: COW<DB::IColumn>::immutable_ptr<DB::IColumn> DB::ColumnVector<unsigned long>::indexImpl<char8_t>(DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> const&, unsigned long) const @ 0x1a4b9e53 in /workspace/clickhouse
2021.02.09 13:44:26.114313 [ 168 ] {} <Fatal> BaseDaemon: 10. ./obj-x86_64-linux-gnu/../src/Columns/ColumnsCommon.h:61: COW<DB::IColumn>::immutable_ptr<DB::IColumn> DB::selectIndexImpl<DB::ColumnVector<unsigned long> >(DB::ColumnVector<unsigned long> const&, DB::IColumn const&, unsigned long) @ 0x1a4743ec in /workspace/clickhouse
2021.02.09 13:44:26.161439 [ 168 ] {} <Fatal> BaseDaemon: 11. ./obj-x86_64-linux-gnu/../src/Columns/ColumnVector.cpp:443: DB::ColumnVector<unsigned long>::index(DB::IColumn const&, unsigned long) const @ 0x1a4742a3 in /workspace/clickhouse
2021.02.09 13:44:26.207524 [ 168 ] {} <Fatal> BaseDaemon: 12. ./obj-x86_64-linux-gnu/../src/Columns/ColumnNullable.cpp:222: DB::ColumnNullable::index(DB::IColumn const&, unsigned long) const @ 0x1a42ef66 in /workspace/clickhouse
2021.02.09 13:44:26.253724 [ 168 ] {} <Fatal> BaseDaemon: 13. ./obj-x86_64-linux-gnu/../src/Columns/ColumnLowCardinality.cpp:852: DB::ColumnLowCardinality::Dictionary::compact(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&) @ 0x1a410aee in /workspace/clickhouse
2021.02.09 13:44:26.324263 [ 168 ] {} <Fatal> BaseDaemon: 14. ./obj-x86_64-linux-gnu/../src/Columns/ColumnLowCardinality.cpp:495: DB::ColumnLowCardinality::compactInplace() @ 0x1a410962 in /workspace/clickhouse
2021.02.09 13:44:26.400330 [ 168 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../src/Columns/ColumnLowCardinality.cpp:487: DB::ColumnLowCardinality::cutAndCompact(unsigned long, unsigned long) const @ 0x1a410880 in /workspace/clickhouse
2021.02.09 13:44:26.463750 [ 168 ] {} <Fatal> BaseDaemon: 16. ./obj-x86_64-linux-gnu/../src/DataTypes/DataTypeLowCardinality.cpp:524: DB::DataTypeLowCardinality::serializeBinaryBulkWithMultipleStreamsImpl(DB::IColumn const&, unsigned long, unsigned long, DB::IDataType::SerializeBinaryBulkSettings&, std::__1::shared_ptr<DB::IDataType::SerializeBinaryBulkState>&) const @ 0x196db6ff in /workspace/clickhouse
2021.02.09 13:44:26.528272 [ 168 ] {} <Fatal> BaseDaemon: 17. ./obj-x86_64-linux-gnu/../src/DataTypes/IDataType.cpp:287: DB::IDataType::serializeBinaryBulkWithMultipleStreams(DB::IColumn const&, unsigned long, unsigned long, DB::IDataType::SerializeBinaryBulkSettings&, std::__1::shared_ptr<DB::IDataType::SerializeBinaryBulkState>&) const @ 0x197a05e0 in /workspace/clickhouse
2021.02.09 13:44:26.598940 [ 168 ] {} <Fatal> BaseDaemon: 18. ./obj-x86_64-linux-gnu/../src/DataStreams/NativeBlockOutputStream.cpp:58: DB::NativeBlockOutputStream::writeData(DB::IDataType const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn> const&, DB::WriteBuffer&, unsigned long, unsigned long) @ 0x1a0d11fc in /workspace/clickhouse
2021.02.09 13:44:26.669783 [ 168 ] {} <Fatal> BaseDaemon: 19. ./obj-x86_64-linux-gnu/../src/DataStreams/NativeBlockOutputStream.cpp:124: DB::NativeBlockOutputStream::write(DB::Block const&) @ 0x1a0d1647 in /workspace/clickhouse
2021.02.09 13:44:26.758379 [ 168 ] {} <Fatal> BaseDaemon: 20. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:1364: DB::TCPHandler::sendData(DB::Block const&) @ 0x1ac7a0cc in /workspace/clickhouse
2021.02.09 13:44:26.848316 [ 168 ] {} <Fatal> BaseDaemon: 21. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:643: DB::TCPHandler::processOrdinaryQueryWithProcessors() @ 0x1ac7898d in /workspace/clickhouse
2021.02.09 13:44:26.937753 [ 168 ] {} <Fatal> BaseDaemon: 22. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:284: DB::TCPHandler::runImpl() @ 0x1ac728e8 in /workspace/clickhouse
2021.02.09 13:44:27.028547 [ 168 ] {} <Fatal> BaseDaemon: 23. ./obj-x86_64-linux-gnu/../src/Server/TCPHandler.cpp:1459: DB::TCPHandler::run() @ 0x1ac7ea78 in /workspace/clickhouse
2021.02.09 13:44:27.132527 [ 168 ] {} <Fatal> BaseDaemon: 24. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x1e4fb39c in /workspace/clickhouse
2021.02.09 13:44:27.202688 [ 168 ] {} <Fatal> BaseDaemon: 25. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:113: Poco::Net::TCPServerDispatcher::run() @ 0x1e4fbba0 in /workspace/clickhouse
2021.02.09 13:44:27.265082 [ 168 ] {} <Fatal> BaseDaemon: 26. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:199: Poco::PooledThread::run() @ 0x1e638d13 in /workspace/clickhouse
2021.02.09 13:44:27.327181 [ 168 ] {} <Fatal> BaseDaemon: 27. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:56: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x1e635b9d in /workspace/clickhouse
2021.02.09 13:44:27.389060 [ 168 ] {} <Fatal> BaseDaemon: 28. ./obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345: Poco::ThreadImpl::runnableEntry(void*) @ 0x1e634a28 in /workspace/clickhouse
2021.02.09 13:44:27.389367 [ 168 ] {} <Fatal> BaseDaemon: 29. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
2021.02.09 13:44:27.389567 [ 168 ] {} <Fatal> BaseDaemon: 30. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
2021.02.09 13:44:28.451829 [ 168 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 832F331D56D43C484FC29CB29BEB2C38. There is no information about the reference checksum.
2021.02.09 13:44:37.592963 [ 82 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

### LINKED ISSUE #23351: Join `Logical error: 'Bad cast from type DB::ColumnLowCardinality to DB::ColumnString`
**Describe the bug**
`Logical error: 'Bad cast from type DB::ColumnLowCardinality to DB::ColumnString'.` in JOIN

https://clickhouse-test-reports.s3.yandex.net/23237/67e427c13aa2b9857c697726251d1e27af4fa1d9/fuzzer_debug/report.html#fail1

```
DROP TABLE IF EXISTS l_lc;
DROP TABLE IF EXISTS r;

CREATE TABLE l_lc (`x` UInt32, `lc` LowCardinality(String)) ENGINE = Memory;
CREATE TABLE r (x UInt32, lc String) ENGINE = Memory;

INSERT INTO r VALUES (0, 'str');
INSERT INTO l_lc VALUES (0, 'str');

SELECT r.lc, toTypeName(l.lc), materialize(r.lc), toTypeName(materialize(r.lc)) FROM l_lc AS l RIGHT JOIN r USING (lc);
```

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Column cardinality in join output same as at the input, close #23351, close #20315

### CODE REVIEW

**Comment by vdimir** [comment]
@mergifyio update

**Comment by vdimir** [comment]
`AST fuzzer` failures is related, but I can't reproduce it locally

**Comment by vdimir** [comment]
I suppose it will be ok after merging fix https://github.com/ClickHouse/ClickHouse/pull/25321

**Review by Avogar — APPROVED** [review]

File: src/Interpreters/join_common.cpp [hunk 1]
```diff
@@ -107,6 +124,19 @@ void convertColumnsToNullable(Block & block, size_t starting_pos)
 /// @warning It assumes that every NULL has default value in nested column (or it does not matter)
 void removeColumnNullability(ColumnWithTypeAndName & column)
 {
+    if (column.type->lowCardinality())
+    {
+        /// LowCardinality(Nullable(T)) case
+        const auto & dict_type = typeid_cast<const DataTypeLowCardinality *>(column.type.get())->getDictionaryType();
+        column.type = std::make_shared<DataTypeLowCardinality>(removeNullable(dict_type));
+
+        ColumnLowCardinality * col_as_lc = typeid_cast<ColumnLowCardinality *>(column.column->assumeMutable().get());
```
Comment (vdimir) [inline src/Interpreters/join_common.cpp:133]: Forgot to change `assumeMutable` to `IColumn::mutate` here, like in `convertColumnToNullable`. I'll create new PR for it now.

**Pull request merged**